libs/ledger/src/consensus/stake_manager.cpp

```diff
@@ -110,30 +162,44 @@ bool StakeManager::ShouldGenerateBlock(Block const &previous, Address const &add
 
 StakeManager::CommitteePtr StakeManager::GetCommittee(Block const &previous)
 {
-  assert(static_cast<bool>(current_));
+  // Calculate the last relevant snapshot
+  uint64_t const last_snapshot =
+      previous.body.block_number - (previous.body.block_number % snapshot_validity_periodicity_);
+
+  // Invalid to request a committee too far ahead in time
+  assert(committee_history_.find(last_snapshot) != committee_history_.end());
 
-  // generate the entropy for the previous block
-  uint64_t entropy{0};
-  if (!LookupEntropy(previous, entropy))
+  if (committee_history_.find(last_snapshot) == committee_history_.end())
   {
-    FETCH_LOG_WARN(LOGGING_NAME, "Unable to lookup committee for ", previous.body.block_number,
-                   " (entropy not ready)");
-    return {};
+    FETCH_LOG_INFO(LOGGING_NAME,
+                   "No committee history found for block: ", previous.body.block_number, " AKA ",
+                   last_snapshot);
   }
 
-  // lookup the snapshot associated
-  auto snapshot = LookupStakeSnapshot(previous.body.block_number);
-  if (!snapshot)
+  FETCH_LOG_INFO(LOGGING_NAME, "Periodicity: ", snapshot_validity_periodicity_);
+
+  // If the last committee was the valid committee, return this. Otherwise, deterministically
+  // shuffle the committee using the random beacon
+  if (last_snapshot == previous.body.block_number)
   {
-    FETCH_LOG_WARN(LOGGING_NAME, "Unable to lookup committee for ", previous.body.block_number);
-    return {};
+    return committee_history_.at(last_snapshot);
   }
+  else
+  {
+    CommitteePtr committee_ptr = committee_history_[last_snapshot];
+    assert(!committee_ptr->empty());
+
+    Committee committee_copy = *committee_ptr;
+
+    DeterministicShuffle(committee_copy, previous.body.random_beacon);
+
+    FETCH_LOG_INFO(LOGGING_NAME, "Rasdfasdf*****");
 
-  // TODO(EJF): Committees can be directly cached
-  return snapshot->BuildCommittee(entropy, committee_size_);
+    return std::make_shared<Committee>(committee_copy);
```

Comment:
I am not sure if I understand the code code correctly. It seems that right now the committee always consists of the same members with random order? Let me know if I misunderstand this.

Reply:
+/- coming and going members.

Reply:
commented in slack